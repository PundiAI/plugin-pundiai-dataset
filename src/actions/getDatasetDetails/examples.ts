export default [
    [
        {
            user: "{{user1}}",
            content: {
                text: "I want to know the details of the dataset with token id 17764 and contract address 0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE.",
            },
        },
        {
            user: "{{agent}}",
            content: {
                text: "Fetching dataset details...",
                action: "GET_DATASET_DETAILS",
            },
        },
    ],

    [
        {
            user: "{{user1}}",
            content: {
                text: "Get dataset 17764 at 0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
            },
        },
        {
            user: "{{agent}}",
            content: {
                text: "Retrieving data...",
                action: "GET_DATASET_DETAILS",
            },
        },
    ],

    [
        {
            user: "{{user1}}",
            content: {
                text: "Show me token 17764",
            },
        },
        {
            user: "{{agent}}",
            content: {
                text: "Fetching dataset (missing contract address)...",
                action: "GET_DATASET_DETAILS",
            },
        },
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "contract Address 0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
            },
        },
        {
            user: "{{agent}}",
            content: {
                text: "Fetching dataset (missing contract address)...",
                action: "GET_DATASET_DETAILS",
            },
        },
    ],
];