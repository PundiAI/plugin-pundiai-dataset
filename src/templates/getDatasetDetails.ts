export const getDatasetDetailsTemplate = `
## Purpose
Extract dataset details from user queries with automatic validation and friendly guidance.
Given the most recent message only, extract information needed to fetch dataset details.

## Dataset Details Structure
A complete dataset details object contains:
{
  "contractAddress": "string (Ethereum address)",
  "tokenId": "string",
  "ownerAddress": "string (Ethereum address)",
  "datasetName": "string (repository/name format)",
  "permissions": "string (public/private)",
  "hash": "string (file hash)",
  "fileSize": "number (bytes)",
  "updateDt": "number (timestamp)",
  "createDt": "number (timestamp)",
  "fileFormatList": ["array of strings"],
  "taskTypeList": ["array of strings"],
  "languageList": ["array of strings"],
  "dataTypeList": ["array of strings"],
  "license": "string",
  "source": "string (e.g. huggingface)",
  "numberOfRows": "number"
}

## Extraction Rules
1. For basic queries, only extract contractAddress and tokenId
2. If user requests full details, extract all available fields
3. Return null for any missing fields

## Examples
1. For "Get dataset 17764 at 0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE":
\`\`\`json
{
  "contractAddress": "0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
  "tokenId": "17764"
}
\`\`\`

2. For full details request:
\`\`\`json
{
  "contractAddress": "0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
  "tokenId": "17764",
  "ownerAddress": "0x5d43dEDC405183DAD609c27BF7e7B0f2225A2B22",
  "datasetName": "May.6.George",
  "permissions": "public",
  "hash": "",
  "fileSize": 0,
  "updateDt": 1746498788217,
  "createDt": 1746498788217,
  "fileFormatList": [],
  "taskTypeList": [],
  "languageList": [],
  "dataTypeList": [],
  "license": "other",
  "source": "",
  "numberOfRows": 0
}
\`\`\`

{{recentMessages}}
Analyze the LAST message only and extract dataset details based on:
1. The specificity of the request (basic or full details)
2. The available information in the message
3. Return only the fields that can be determined
4. Provide null for missing fields with guidance when appropriate`;