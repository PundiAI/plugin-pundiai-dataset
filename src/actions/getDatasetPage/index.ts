import { Action, HandlerCallback, IAgentRuntime, Memory, State, elizaLogger, composeContext, ModelClass, generateObjectDeprecated } from "@elizaos/core";
import examples from "./examples.ts";
import { getV1ViewQueryDatasetPage } from "../../utils/service"
import { GetV1ViewQueryDatasetPageQuery } from "../../utils/service/index.type.ts"
import { getDatasetPageTemplate } from "../../templates/getDatasetPage.ts"

export const getDatasetPage: Action = {
    name: "GET_DATASET_PAGE",
    description: "Get information about the dataset",
    similes: [],
    examples: examples,
    validate: async (_runtime: IAgentRuntime, message: Memory) => {
        return true
    },
    handler: async (runtime: IAgentRuntime, message: Memory, state?: State, _options?: { [key: string]: unknown; }, callback?: HandlerCallback): Promise<boolean> => {
        elizaLogger.log("Starting Moralis GET_DATASET_PAGE handler...");

        let currentState: State;
        if (!state) {
            currentState = (await runtime.composeState(message)) as State;
        } else {
            currentState = await runtime.updateRecentMessageState(state);
        }

        try {
            const statsContext = composeContext({
                state: currentState,
                template: getDatasetPageTemplate,
            });
            const content = (await generateObjectDeprecated({
                runtime,
                context: statsContext,
                modelClass: ModelClass.LARGE,
            })) as unknown as GetV1ViewQueryDatasetPageQuery;

            if (!content || typeof content !== "object") {
                throw new Error("Invalid response format from model");
            }

            const params = {
                ...(content.datasetName && { datasetName: content.datasetName }),
                ...(content.address && { address: content.address }),
                ...(content.addressType && { 
                  addressType: Number(content.addressType) 
                }),
                ...(content.datasetDataType && { 
                  datasetDataType: content.datasetDataType 
                }),
                ...(content.datasetFormat && { 
                  datasetFormat: content.datasetFormat 
                }),
                ...(content.permissions && { 
                  permissions: content.permissions 
                }),
                ...(content.fileSizeStart && { 
                    fileSizeStart: Number(content.fileSizeStart)
                  }),
                  ...(content.fileSizeEnd && { 
                    fileSizeEnd: Number(content.fileSizeEnd)
                  }),
                pageIndex: 1, 
                pageSize: 20,
              };

            elizaLogger.log("Request params:", params);

            const response = await getV1ViewQueryDatasetPage(params)

            const datasetData = response.data;

            if (callback) {

                if (datasetData.data.length <= 0){
                    callback({
                        text: "Here is the dataset not data.",
                        content: {
                            params,
                            datasetData
                        },
                    });
                    return true;
                }

                callback({
                    text: "Here is the dataset data: \n" + JSON.stringify(datasetData, null, 2),
                    content: {
                        params,
                        datasetData
                    },
                });
            }

            return true;
        } catch (error: unknown) {
            elizaLogger.error("Error in GET_DATASET_PAGE handler:", error);
            const errorMessage =
                error instanceof Error ? error.message : String(error);
            if (callback) {
                callback({
                    text: `Error fetching dataset data: ${errorMessage}`,
                    content: { error: errorMessage },
                });
            }

            return false;
        }
    }
}

