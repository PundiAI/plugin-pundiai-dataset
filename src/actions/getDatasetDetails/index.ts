import { Action, HandlerCallback, IAgentRuntime, Memory, State, elizaLogger, composeContext, ModelClass, generateObjectDeprecated } from "@elizaos/core";
import examples from "./examples.ts";
import { getV1ViewQueryDatasetDetails } from "../../utils/service"
import { GetV1ViewQueryDatasetDetailsQuery } from "../../utils/service/index.type.ts"
import { getDatasetDetailsTemplate } from "../../templates/getDatasetDetails";

export const getDatasetDetails: Action = {
    name: "GET_DATASET_DETAILS",
    description: "Get data set details by token id and contract address",
    similes: [],
    examples: examples,
    validate: async (_runtime: IAgentRuntime, message: Memory) => {
        return true
    },
    handler: async (runtime: IAgentRuntime, message: Memory, state?: State, _options?: { [key: string]: unknown; }, callback?: HandlerCallback): Promise<boolean> => {
        elizaLogger.log("Starting Moralis GET_DATASET_DETAILS handler...");

        let currentState: State;
        if (!state) {
            currentState = (await runtime.composeState(message)) as State;
        } else {
            currentState = await runtime.updateRecentMessageState(state);
        }

        try {
            const statsContext = composeContext({
                state: currentState,
                template: getDatasetDetailsTemplate,
            });
            const content = (await generateObjectDeprecated({
                runtime,
                context: statsContext,
                modelClass: ModelClass.LARGE,
            })) as unknown as GetV1ViewQueryDatasetDetailsQuery;

            if (!content || typeof content !== "object") {
                throw new Error("Invalid response format from model");
            }

            if (!content.tokenId || content.tokenId === 'null') {
                throw new Error("Please provide the Token ID for this contract");
            }

            if (!content.contractAddress || content.contractAddress === 'null') {
                throw new Error("Please provide the contract address");
            }

            const params = {
                tokenId: content.tokenId ,
                contractAddress: content.contractAddress
            };

            elizaLogger.log("Request params:", params);

            const response = await getV1ViewQueryDatasetDetails(params)

            const detailsData = response.data;

            if (callback) {
                callback({
                    text: "Here are the dataset details: \n" + JSON.stringify(detailsData, null, 2),
                    content: {
                        params,
                        detailsData
                    },
                });
            }

            return true;
        } catch (error: unknown) {
            elizaLogger.error("Error in GET_DATASET_DETAILS handler:", error);
            const errorMessage =
                error instanceof Error ? error.message : String(error);
            if (callback) {
                callback({
                    text: `Error fetching dataset details: ${errorMessage}`,
                    content: { error: errorMessage },
                });
            }

            return false;
        }
    }
}

