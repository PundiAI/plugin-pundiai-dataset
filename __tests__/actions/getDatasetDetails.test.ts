import { describe, it, expect, vi, beforeEach, } from 'vitest';
import { elizaLogger, ModelClass, generateObjectDeprecated, composeContext } from '@elizaos/core';
import { getDatasetDetails } from "../../src/actions/getDatasetDetails";

describe('getDatasetDetails action', () => {
    const mockRuntime = {
        composeState: vi.fn(),
        updateRecentMessageState: vi.fn(),
        getPluginConfig: vi.fn(),
    };

    const mockMessage = {};
    const mockState = {};
    const mockCallback = vi.fn();

    beforeEach(() => {
        vi.clearAllMocks();

        // Mock runtime functions
        mockRuntime.composeState.mockResolvedValue(mockState);
        mockRuntime.updateRecentMessageState.mockResolvedValue(mockState);
        mockRuntime.getPluginConfig.mockResolvedValue({});

        // Mock the core functions
        vi.mocked(elizaLogger.log).mockImplementation(() => {});
        vi.mocked(elizaLogger.error).mockImplementation(() => {});
        vi.mocked(elizaLogger.success).mockImplementation(() => {});
        vi.mocked(composeContext).mockReturnValue({});
    });

    it('should fetch and format details data', async () => {

        vi.mocked(generateObjectDeprecated).mockResolvedValueOnce({
            tokenId: '17764',
            contractAddress: '0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE',
          });

        await getDatasetDetails.handler(mockRuntime, mockMessage, mockState, {}, mockCallback);

        expect(mockCallback).toMatchSnapshot();
    })
});