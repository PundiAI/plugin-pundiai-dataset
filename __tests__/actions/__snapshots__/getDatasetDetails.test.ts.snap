// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDatasetDetails action > should fetch and format details data 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "content": {
          "detailsData": {
            "contractAddress": "0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
            "createDt": 1746498788217,
            "dataTypeList": [],
            "datasetName": "May.6.George",
            "fileFormatList": [],
            "fileSize": 0,
            "hash": "",
            "languageList": [],
            "license": "other",
            "numberOfRows": 0,
            "ownerAddress": "0x5d43dEDC405183DAD609c27BF7e7B0f2225A2B22",
            "permissions": "public",
            "source": "",
            "taskTypeList": [],
            "tokenId": "17764",
            "updateDt": 1746498788217,
          },
          "params": {
            "contractAddress": "0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
            "tokenId": "17764",
          },
        },
        "text": "Here are the dataset details: 
{
  "contractAddress": "0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
  "tokenId": "17764",
  "ownerAddress": "0x5d43dEDC405183DAD609c27BF7e7B0f2225A2B22",
  "datasetName": "May.6.George",
  "permissions": "public",
  "hash": "",
  "fileSize": 0,
  "updateDt": 1746498788217,
  "createDt": 1746498788217,
  "fileFormatList": [],
  "taskTypeList": [],
  "languageList": [],
  "dataTypeList": [],
  "license": "other",
  "source": "",
  "numberOfRows": 0
}",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
